Source/DafnyStandardLibraries/examples/Parsers/Tutorial.dfy

```diff
@@ -0,0 +1,109 @@
+/*******************************************************************************
+ *  Copyright by the contributors to the Dafny Project
+ *  SPDX-License-Identifier: MIT 
+ *******************************************************************************/
+
+module ExampleParsers.Tutorial {
+  import opened Std.Parsers.StringParsers
+
+  method {:test} TestSplit1() {
+    var nonComma: Parser<string> :=
+      OneOrMore(CharTest((c: char) => c != ',', "non-comma"));
+    var p :=
+      Bind(
+        nonComma, (result: string) =>
+          Rep(ConcatKeepRight(String(","), nonComma),
+              (acc, elem) => acc + [elem],
+              [result]
+          ));
+    var input := "abc,d,efg";
+    var result := p(ToInput(input));
+    expect result.ParseSuccess?;
+    expect result.result == ["abc","d","efg"];
+    var source := "abc,d,,";
+    expect p(ToInput(source))
+        == ParseResult.ParseFailure(Recoverable, FailureData("expected a non-comma", A.Input(source, |source| - 1, |source|), Option.None));
+    var failureMessage := FailureToString(source, p(ToInput(source)));
+    expect failureMessage
+        == "Error:"               + "\n" +
+           "1: abc,d,,"           + "\n" +
+           "         ^"           + "\n" +
+           "expected a non-comma" + "\n";
+  }
+
+  function flatten<A>(): ((A, (A, A))) -> (A, A, A) {
+    (input: (A, (A, A))) =>
+      (input.0, input.1.0, input.1.1)
+  }
+
+  method {:test} TestTicTacToe() {
+    var x := OrSeq(
+      [
+        String("O"), String("X"), String(" ")
+      ]);
+    var v := String("|");
+    var row := Map(Concat(x, ConcatKeepRight(v, Concat(x, ConcatKeepRight(v, x)))),
+                   flatten<string>());
+    var sep := String("\n-+-+-\n");
+    var grid := Map(
+      Concat(row, ConcatKeepRight(sep, Concat(row, ConcatKeepRight(sep, row)))),
+      flatten<(string, string, string)>())
+      ;
+    var input := ToInput("O|X| \n-+-+-\nX|O| \n-+-+-\nP| |O");
+    // 012345 678901 234567 890123 45678
+    var r := grid(input);
+    expect r.IsFailure();
+    expect A.Length(input) - A.Length(r.data.remaining) == 24;
+    expect r.data.message == "expected 'O'";
+    expect r.data.next.Some?;
+    expect r.data.next.value.message == "expected 'X'";
+    expect r.data.next.value.next.Some?;
+    expect r.data.next.value.next.value.message == "expected ' '";
+    expect r.data.next.value.next.value.next.None?;
```

Comment:
(nit) Couldn't this be `expect r == ParseSuccess(...)`?

Reply:
Since the failure message is already tested below and the test is more readable, I'm removing these intermediate tests so that we don't test the internal representation.